x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexer.java

```diff
@@ -252,7 +257,12 @@ public synchronized boolean maybeTriggerAsyncJob(long now) {
     protected abstract void onAbort();
 
     private void finishWithFailure(Exception exc) {
-        doSaveState(finishAndSetState(), position.get(), () -> onFailure(exc));
+        try {
+            doSaveState(finishAndSetState(), position.get(), () -> onFailure(exc));
```

Comment:
`doSaveState` is async so there is no need to try/catch this call. We ignore exceptions in `doSaveState`, this is ok IMO since saving the state here is just to ensure that we'll start from the last commit point.

Reply:
That's true for the rollup implementation (at the moment), but not from a generic Async-2-PhaseIndexer perspective if the implementation of doSaveState(...) is buggy. In such a case error handling is trapped in 2 exception handlers:

1. any error codition, exception is thrown
2. `finishWithFailure` called the 1st time, throws another exception from `doSaveState`
3. exception is handled and calls `finishWithFailure` a 2nd time
4. `finishAndSetState` throws `Indexer job encountered an illegal state [STARTED]` as step 2 was incomplete
5.  no handler for this exception

Just want to make sure the problem is understood (the unit test should also show it). I admit it's somewhat made up a bit, assuming a buggy `doSaveState`.

Alternatively I could just accept the fact that `doSaveState` simply has to be implemented so it never throws, for which case I would be fine with a small doc fix.

Reply:
I don't think we should save the state at the end of the job. In fact the indexer should never save any state, this should be done outside of the task. I have a branch that does that but it's not ready yet. In the mean time I think it's fine to just ignore the exceptions thrown by `doSaveState` and just call `onFailure` with the real one (the one thrown by the search/bulk).